Sources/Datadog/DatadogInternal/Upload/Event.swift

```diff
@@ -0,0 +1,67 @@
+/*
+ * Unless explicitly stated otherwise all files in this repository are licensed under the Apache License Version 2.0.
+ * This product includes software developed at Datadog (https://www.datadoghq.com/).
+ * Copyright 2019-Present Datadog, Inc.
+ */
+
+import Foundation
+
+/// Struct representing a single event.
+public struct Event: Equatable {
+    /// Data representing the event.
+    public let data: Data
+
+    /// Metadata associated with the event.
+    /// Metadata is optional and may be `nil` but of very small size.
+    /// This allows us to skip resource intensive operations in case such
+    /// as filtering of the events.
+    public let metadata: Data?
```

Comment:
nit/ Because we pass metadata in `write<E: Encodable, M: Encodable>(event: E, metadata: M)` the feature doesn't know the actual detail of `metadata: Data` encoding. We should clearly state in this comment that `metadata: Data?` is JSON encoded. Actually, the same comment should be applied to `data: Data` field.

Reply:
Currently, responsibilities are not properly assigned: Core owns the encoding while Features are responsible of decoding. JSON encoding/decoding is only a convention between the core and Features, so I think it should be mentioned in comments.

Ultimately, I think we should transfer responsibility of encoding to the Features, so they are free to use the serialisation of their choices (JSON, TLV, protobuf, ...). Not a change request of this PR but something to consider in the future.